app/models/concerns/twitch_user.rb

```diff
@@ -6,20 +6,22 @@ module TwitchUser
 
   included do
     def sync_twitch_follows!
+      success = false
       ActiveRecord::Base.transaction do
         current_followed_users = User.where(twitch_id: Twitch::Follows.followed_ids(twitch_id))
         old_followed_users = twitch_followed_users
 
-        (current_followed_users - old_followed_users).each do |u|
-          TwitchUserFollow.create!(from_user: self, to_user: u)
-        end
+        TwitchUserFollow.import((current_followed_users - old_followed_users).map do |u|
+          TwitchUserFollow.new(from_user: self, to_user: u)
+        end)
 
         (old_followed_users - current_followed_users).each do |u|
           TwitchUserFollow.find(from_user: self, to_user: u).destroy
         end
 
-        update(twitch_user_follows_checked_at: Time.now.utc)
+        success = true
       end
+      update(twitch_user_follows_checked_at: nil) unless success
```

Comment:
Ah, I see what you were going for now with updating this last, since we don't want to update it in the case of failure. In that case we may want to use [locks][1]:

```ruby
with_lock do
  # fetch follows
  # update twitch_user_follows_checked_at
  # no worrying about if someone else will update it out from under you
end
```

[1]: http://api.rubyonrails.org/classes/ActiveRecord/Locking/Pessimistic.html

Reply:
My concern with this is that jobs are not guaranteed to be run instantly, so there is a case that if we get an extremely long request time from Twitch then it might queue up many different jobs to do the same thing if the user hits the home page (or other pages that request the sync in the future) a couple of times.  My initial thought was to try and keep all time based logic out of the method since it should only deal with getting and updating the followers.  If you aren't concerned with queuing jobs a lot we could just queue every home page load and have the time check in the `sync_follows` method, or I could add an additional check at the beginning of the method as well and have multiple checks.

The other concern with locking the user is it might lock up other actions that need to update the user (logging back in) if the response from Twitch is slow or we get a ridiculous timeout.